service/storage_service.cc

```diff
@@ -5649,6 +5649,63 @@ future<> storage_service::move_tablet(table_id table, dht::token token, locator:
         return !tmap.get_tablet_transition_info(tmap.get_tablet_id(token));
     });
 }
+}
+// TODO: remove this header. It's currently needed to throw cql3 exception if topology_global_queue_empty() == true
+#include "cql3/cql_statement.hh"
+namespace service {
+    future<> storage_service::alter_tablets_keyspace(sstring ks_name, std::map<sstring, sstring> replication_options,
+                                                     std::optional<service::group0_guard>& guard) {
+        utils::UUID request_id;
+        if (this_shard_id() != 0) {
+            rtlogger.info("if (this_shard_id() != 0)");
+            co_return;
+        }
+
+        bool had_guard = guard.has_value();
+
+        while (true) {
+            rtlogger.info("while (true)");
+            if (topology_global_queue_empty()) {
+                rtlogger.info("if (topology_global_queue_empty())");
+                if (not guard.has_value())
+                    guard = co_await _group0->client().start_operation(&_group0_as);
+                rtlogger.info("acquiring guard done");
+
+                topology_mutation_builder builder(guard->write_timestamp());
+                builder.set_global_topology_request(global_topology_request::keyspace_rf_change);
+                builder.set_new_keyspace_rf_change_data(ks_name, replication_options);
+                topology_change change{{builder.build()}};
+                rtlogger.info("building topology_change done");
+
+                group0_command g0_cmd = _group0->client().prepare_command(std::move(change), *guard, "alter_tablets_keyspace");
+                request_id = guard->new_group0_state_id();
+                rtlogger.info("prepare_command done");
+                try {
+                    co_await _group0->client().add_entry(std::move(g0_cmd), std::move(*guard), &_abort_source);
+                    rtlogger.info("co_await _group0->client().add_entry done");
+
+                    rtlogger.info("before co_await wait_for_topology_request_completion");
+                    // TODO: fix waiting for global topology req
+//                    auto error = co_await wait_for_topology_request_completion(request_id);
+//                    rtlogger.info("co_await wait_for_topology_request_completion done");
+//
+//                    if (!error.empty()) {
+//                        throw std::runtime_error(fmt::format("alter_tablets_keyspace failed with: {}", error));
+//                    }
+
+                    if (had_guard)
+                        guard = co_await _group0->client().start_operation(&_group0_as);
```

Comment:
That seems wrong, it breaks transaction atomicity. If we don't need atomicity, there should be no need to give the guard back.

Reply:
I think it relates to this TODO comment - https://github.com/scylladb/scylladb/pull/16723/files#diff-bb7d446526823f40f72ff83ac80f5b1dd95ca6019e0342d222aed963eb4e8a23R126-R130
If merging both topology mutations and schema mutations is a viable option to resolve this problem, I'll perform all these mutations under a single raft guard.